docs/contributing/999-rfcs/2024-03-14-explicit-lift-qualification.md

```diff
@@ -0,0 +1,198 @@
+}
+```
+3. Cases where our knowledge of the application logic can produce better qualifications:
+```js
+for i in buckets.len {
+  if i % 2 == 0 {
+    buckets.at(i).put("key", "value"); // We only need to qualify the even buckets with "put" and not the odd ones
+  }
+}
+```
+4. Cases where just the method name isn't enough information to create least privilege permissions:
+```js
+for i in 0..5 {
+  bucket.put("k{i}", "value");
+}
+```
+Here we can benefit from an explicit qualification of `bucket` with `"put"` operation and a specific key pattern: `"k*"`. We currently don't
+have a way to disable the automatic qualification and to define the narrower qualification.
+
+Note that in all the above exmaples except the last one we currently get a compilation error due to us trying to use a `Bucket.put` operation on an inflight expression.
+
+## Requirements
+
+So the aim of this proposal is to present:
+1. A way to explicitly qualifying the preflight objects that are being used by inflight methods/closures.
+2. Disabling compilation errors where the explicit qualification apply.
+3. Disabling automatic qualification where the explicit qualification apply. 
+4. Provide a framework that can be extended to programatically defining and reusing lift qualifications.
+
+## Naming
+
+"qualifiations" is used because it qualifies how the preflight object is being used. 
+Other words that we sometime used in the past and might be clearer are "capabilities" or "operations":
+* "capabilities" makes sense because it describes the capabilities the host needs in order to successfully run the inflight code.
+* "operations" makes sense because it describes the operations being performed on the preflight objects by the inflight code.
+
+In this proposal I'll use "qualifications", but I'd be happy to hear what you think.
+
+## Prior work
+### Issue with discussion on the subject: 
+[Mechanism for explicitly qualifying lifting](https://github.com/winglang/wing/issues/76)
+### Existing PR
+[feat(compiler): allow explicit lift qualifications of preflight objects](https://github.com/winglang/wing/pull/5935)
+
+Issues with this PR:
+* Confusing sytax that looks like an inflight function call might lead to [misuse](https://github.com/winglang/wing/pull/5935#issuecomment-1995051115).
+* No scoping of where we're disabling compiler errors auto qualification.
+* No mechanism or clear path to mechanism for programatically defining and reusing lift qualifications.
+
+## Proposal
+
+```js
+let b1 = new cloud.Bucket() as "b1";
+let b2 = new cloud.Bucket() as "b2";
+inflight () => {
+  let var b = b1;
+  if true {
+    b = b2;
+  }
+  // In the following scope we disable qualification compilation errors due to unknown preflight object
+  // and we also disable automatic qualification.
+  // We explicitly qualify b1 and b2 with "put" since the compiler doesn't figure this out for us.
+  with lift_qualification {b1: ["put"], b2: ["put"]} {
```

Comment:
Some ideas other ideas for the naming could be `with capability <C> { ... }` or `using qualification <Q> { ... }`

Reply:
Do we have an idea about what `with` or `using` is actually doing here from a user's perspective? Just the keyword I mean, not the entire construct.

If we don't have that fleshed out, I feel like we should just keep it as short as possible and avoid attributing any sort of "kind" like `lift_qualification`. i.e. just:

```js
with { b1: ["put"], b2: ["put"] } {
  // ...
}

let quals = new std.Qualifications();
with quals {
  // ...
}
```

or even bring back lift:

```js
lift { b1: ["put"], b2: ["put"] } {
  // ...
}
```

Reply:
> Do we have an idea about what with or using is actually
I kind of like python's [context manager pattern](https://docs.python.org/3/reference/datamodel.html#with-statement-context-managers) for generalizing `with`, but sadly we're dealing with "comp-time" here. If you have any idea how to reconcile this and add a "context manager" to wing that'll work here, then `with` seems like good syntax. If not, then I agree, lets go with `lift { stuff } { .. }`.

Reply:
+1 on `lift <map> { ... }`: